concordium-consensus/src/Concordium/TransactionVerification.hs

```diff
@@ -0,0 +1,253 @@
+      Just ipInfo ->
+        case credential of
+          ID.InitialACWP icdi ->
+            -- check signatures for an initial credential deployment
+            unless (A.verifyInitialAccountCreation ipInfo messageExpiry (S.encode icdi)) $ throwError CredentialDeploymentInvalidSignatures
+          ID.NormalACWP ncdi -> do
+            cryptoParams <- lift getCryptographicParameters
+            let ncdv = ID.cdiValues ncdi
+            mArsInfos <- lift (getAnonymityRevokers (OrdMap.keys (ID.cdvArData ncdv)))
+            case mArsInfos of
+              -- check that the anonymity revokers exist
+              Nothing -> throwError CredentialDeploymentInvalidAnonymityRevokers
+              Just arsInfos -> do
+                -- if the credential deployment contained an empty map of 'ChainArData' then the result will be 'Just empty'.
+                when (null arsInfos) $ throwError CredentialDeploymentInvalidAnonymityRevokers
+                -- check signatures for a normal credential deployment
+                unless (A.verifyCredential cryptoParams ipInfo arsInfos (S.encode ncdi) (Left messageExpiry)) $ throwError CredentialDeploymentInvalidSignatures
+    return CredentialDeploymentSuccess)
+
+-- |Verifies a 'ChainUpdate' transaction.
+-- This function verifies the following:
+-- * Checks that the effective time is no later than the timeout of the chain update.
+-- * Checks that sequence number is ok.
+-- * Checks that the 'ChainUpdate' is correctly signed.
+verifyChainUpdate :: TransactionVerifier m => ST.UpdateInstruction -> m VerificationResult
+verifyChainUpdate ui@ST.UpdateInstruction{..} =
+  either id id <$> runExceptT (do
+    -- check that the effective time is not after the timeout of the chain update.
+    when (ST.updateTimeout uiHeader >= ST.updateEffectiveTime uiHeader && ST.updateEffectiveTime uiHeader /= 0) $
+      throwError ChainUpdateEffectiveTimeBeforeTimeout
+    -- check that the sequence number is ok
+    nextSN <- lift $ getNextUpdateSequenceNumber (ST.updateType (ST.uiPayload ui))
+    let nonce = ST.updateSeqNumber (ST.uiHeader ui)
+    unless (nextSN <= nonce) $ throwError ChainUpdateInvalidSequenceNumber
+    -- check the signature is valid
+    keys <- lift getUpdateKeysCollection
+    unless (Updates.checkAuthorizedUpdate keys ui) $ throwError ChainUpdateInvalidSignatures
+    return (ChainUpdateSuccess (getHash keys) nonce))
+
+-- |Verifies a 'NormalTransaction' transaction.
+-- This function verifies the following:
+-- * Checks that enough energy is supplied for the transaction.
+-- * Checks that the sender is a valid account.
+-- * Checks that the nonce is correct.
+-- * Checks that the 'NormalTransaction' is correctly signed.
+--
+-- Note. This function does not assert that the sender account has enough funds to
+-- cover the transfer. This check cannot be cached and as such it must always be carried out
+-- by the 'Scheduler'. 
```

Comment:
I think we should reject transactions when we receive them individually and the sender does not have enough tokens to cover the deposit at the time we receive it. Otherwise we have a loophole. Of course we cannot do this for transactions that are part of a block.